[WK2][macOS] Update sandbox profile caching to not require file locking

cdumez opened the pull request:
#### 9ec2b57dca82a9fafbfd7bf42f243d3e41cfbe1d
<pre>
[WK2][macOS] Update sandbox profile caching to not require file locking
<a href="https://bugs.webkit.org/show_bug.cgi?id=258940">https://bugs.webkit.org/show_bug.cgi?id=258940</a>
rdar://111383802

Reviewed by Ben Nham, Per Arne Vollan and Brent Fulgham.

As a process launch optimization on macOS, we cache the compiled sandbox
profile to disk and reuse this cached profile for follow-up process
launches.

The logic was relying to using file locking to avoid reading partially
written sandbox profiles, which would lead to issues if the process
suspends while in the middle of doing this work.

To avoid locking, we now write the compiled sandbox profile to a
temporary path which includes the current process&apos;s PID, before moving
it to the final destination (using rename()). Because rename() is atomic,
this should achieve what we want without file locking.

* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::writeSandboxDataToCacheFile):
(WebKit::tryApplyCachedSandbox):

Canonical link: <a href="https://commits.webkit.org/265823@main">https://commits.webkit.org/265823@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3ed7d4702376dacdd49fe2fb5dd21332f5f50c20

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11977 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12323 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12626 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13723 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11560 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14733 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12337 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14260 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12141 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12953 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10134 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14135 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10245 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17997 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11323 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11029 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14198 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11512 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9471 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10726 "Built successfully") | | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2933 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15051 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11370 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham reviewed (commented):
> Before rename() is atomic

Because rename() is atomic

I think it is safe to deterministically use the PID in the temporary filename here since we truncate the file temporary file before writing it (in the case that some malicious process tries to write to that same path, for instance).

Might also be a good idea to remove the `lockMode` stuff entirely from `fileContents` since it seems to be unused now.

brentfulgham commented on Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
```diff
@@ -393,8 +393,13 @@ static bool ensureSandboxCacheDirectory(const SandboxInfo& info)
+    // then rename it to the final cache path.
+    auto temporaryPath = makeString(info.filePath, '-', getpid());
+    FileHandle file { temporaryPath, FileSystem::FileOpenMode::Truncate };
+    if (file.write(cacheFile.data(), cacheFile.size()) != safeCast<int>(cacheFile.size()))
```
Should we clean up the temp file? I don't believe it automatically happens.

cdumez replied:
Sounds good. I am now deleting the temporary file in this case.

pvollan reviewed (approved):
R=me.

brentfulgham reviewed (approved):
Excellent!

cdumez committed:
[WK2][macOS] Update sandbox profile caching to not require file locking
https://bugs.webkit.org/show_bug.cgi?id=258940
rdar://111383802

Reviewed by Ben Nham, Per Arne Vollan and Brent Fulgham.

As a process launch optimization on macOS, we cache the compiled sandbox
profile to disk and reuse this cached profile for follow-up process
launches.

The logic was relying to using file locking to avoid reading partially
written sandbox profiles, which would lead to issues if the process
suspends while in the middle of doing this work.

To avoid locking, we now write the compiled sandbox profile to a
temporary path which includes the current process's PID, before moving
it to the final destination (using rename()). Because rename() is atomic,
this should achieve what we want without file locking.

* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::writeSandboxDataToCacheFile):
(WebKit::tryApplyCachedSandbox):

Canonical link: https://commits.webkit.org/265823@main
